src/coreclr/src/jit/lowerxarch.cpp

```diff
@@ -947,6 +947,61 @@ void Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
             return;
         }
 
+        case NI_SSE2_Insert:
+        case NI_SSE41_Insert:
+        case NI_SSE41_X64_Insert:
+        {
+            assert(HWIntrinsicInfo::lookupNumArgs(node) == 3);
+
+            GenTreeArgList* argList = node->gtOp1->AsArgList();
+            GenTree*        op2     = argList->Rest()->Current();
+
+            if (!op2->OperIs(GT_CAST))
+            {
+                break;
+            }
+
+            // Insert takes either a 32-bit register or a memory operand.
+            // In either case, only gtSIMDBaseType bits are read and so
+            // widening or narrowing the operand may be unnecessary and it
+            // can just be used directly.
+
+            GenTree* castOp = op2->AsCast()->CastOp();
+
+            if (genTypeSize(castOp->gtType) >= genTypeSize(node->gtSIMDBaseType))
+            {
+                BlockRange().Remove(op2);
+                argList->Rest()->gtOp1 = castOp;
+            }
+            break;
+        }
+
+        case NI_SSE42_Crc32:
+        {
+            assert(HWIntrinsicInfo::lookupNumArgs(node) == 2);
+
+            GenTree* op2 = node->gtOp2;
```

Comment:
It's not a lot of code, but it's unfortunate to duplicate this. You might consider creating a method that takes the possible cast op, and then returns it or its child.

Reply:
Are you thinking something like:
```csharp
node->gtOp2 = RemoveCastIfPresent(node->gtType, node->gtOp2);
```

and
```csharp
GenTree* RemoveCastIfPresent(var_types expectedType, GenTree* op)
{
    if (!op->OperIs(GT_CAST))
    {
        return op;
    }

    GenTree* castOp = op2->AsCast()->CastOp();

    if (genTypeSize(castOp->gtType) >= genTypeSize(expectedType))
    {
        BlockRange().Remove(op);
        return castOp;
    }

    return op;
}
```

?

Reply:
Maybe `TryRemoveCast` (same signature) would be a better name?

Reply:
Yes, something like that was what I was thinking. As I say, though, it's not a lot of code so I don't want to block you on it.

Reply:
Sounds good! I've basically got it done locally (just doing some sanity testing) so I'll merge this and will get a follow up PR shortly after.

The downside on ARM64 is that the existing use-case isn't easy to share. It is conditionalized on the underlying operation being a constant and being within a specific range. But I imagine there may be other use-cases that can share in the future